ocaml/xapi/message_forwarding.ml

```diff
@@ -746,8 +746,14 @@ functor
           (current_pool_uuid ~__context)
           (host_uuid ~__context host) ;
         let local_fn = Local.Pool.eject ~host in
-        do_op_on ~local_fn ~__context ~host (fun session_id rpc ->
-            Client.Pool.eject rpc session_id host)
+        let hosts = Xapi_pool_helpers.get_master_slaves_list ~__context in
+        (* call eject on all slaves but only slave "host" will eject
+           itself *)
+        hosts
+        |> List.iter (fun h ->
+               do_op_on ~local_fn ~__context ~host:h (fun session_id rpc ->
+                   Client.Pool.eject rpc session_id host)) ;
```

Comment:
Is the order in which you call Pool.eject on all the hosts important?

Suppose that the master-slave list looks like `[ master ; slave_1 ; slave_2 ]`, and that `slave_2` is the ejectee. The master will first remove the ejectee's certificates, then the master will ask `slave_1` to remove the ejectee's certificates, and then finally the master will ask `slave_2` to eject. However that will fail because the master no longer trusts slave_2

It might be safer to do this:

```
Client.Pool.eject rpc session_id ejectee ;
hosts |> List.filter (fun h -> h <> ejectee) |> List.iter (fun h -> Client.Pool.eject rpc session_id) ;
```

Reply:
I agree - the core eject operation needs to happen first before we can remove certs.